File: 2024/day14/solution.go
package main

import (
	adventofcode "adventofcode/utils"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

type Position struct {
	x int
	y int
}

type Bot struct {
	pos  Position
	xVel int
	yVel int
}

func main() {
	lines := getContentAsListOfStrings("./input.txt")

	var bots []Bot

	// for the simple input
	// xLimit, yLimit := 11, 7

	xLimit, yLimit := 101, 103

	for _, line := range lines {
		botX, botY, botXVel, botYVel := getBotFromLine(line)
		bots = append(bots, Bot{Position{botX, botY}, botXVel, botYVel})
	}

	for move := 0; move < 100; move++ {
		for i := range bots {
			moveBot(&bots[i], xLimit, yLimit)
		}
	}

	q1, q2, q3, q4 := countBotsInQuadrants(bots, xLimit, yLimit)
	fmt.Println("PART 1: ", q1*q2*q3*q4)
}

// DEBUG
// func printGrid(bots []Bot, xLimit, yLimit int) {
// 	for y := 0; y < yLimit; y++ {
// 		for x := 0; x < xLimit; x++ {
// 			fmt.Print(countBotPositions(bots, xLimit, yLimit)[x][y], " ")
// 		}
// 	}
// }

func moveBot(bot *Bot, xlim, ylim int) {
	newX := bot.pos.x + bot.xVel

	// if new position is out of bounds, teleport to other side
	if newX < 0 {
		newX = newX + xlim
	} else if newX >= xlim {
		newX = newX - xlim
	}

	newY := bot.pos.y + bot.yVel

	if newY < 0 {
		newY = newY + ylim
	} else if newY >= ylim {
		newY = newY - ylim
	}

	bot.pos = Position{newX, newY}
}

func countBotsInQuadrants(bots []Bot, xLimit, yLimit int) (int, int, int, int) {
	midX, midY := xLimit/2, yLimit/2
	q1, q2, q3, q4 := 0, 0, 0, 0

	for _, bot := range bots {
		x, y := bot.pos.x, bot.pos.y

		// skip robots on the middle lines
		if x == midX || y == midY {
			continue
		}

		// determine quadrant
		if x > midX && y < midY {
			q1++
		} else if x < midX && y < midY {
			q2++
		} else if x < midX && y > midY {
			q3++
		} else if x > midX && y > midY {
			q4++
		}
	}

	return q1, q2, q3, q4
}

func getBotFromLine(line string) (int, int, int, int) {
	// match numbers (optional negative sign, followed by digits)
	re := regexp.MustCompile(`-?\d+`)
	matches := re.FindAllString(line, -1)

	var numbers []int
	for _, match := range matches {
		num, err := strconv.Atoi(match)
		if err == nil {
			numbers = append(numbers, num)
		}
	}

	return numbers[0], numbers[1], numbers[2], numbers[3]
}

func getContentAsListOfStrings(in string) []string {
	content := adventofcode.GetFileContent(in)

	// windows line endings
	return strings.Split(content, "\r\n")
}

File: 2024/day14/solution.ts
import { getFileContent } from "../utils/tsutils";

type Bot = {
  pos: Coordinate;
  velocity: Coordinate;
}

type Coordinate = {
  x: number;
  y: number; 
}

const configuration = {
  seconds: 100,
  xLim: 101,
  yLim: 103
} as const;

// logic
const content = (await getFileContent('./input.txt')).split('\n')

const bots: Bot[] = content.map(line => {
  const coords = getBotFromLine(line)
  return {
    pos: { x: coords.x, y: coords.y },
    velocity: { x: coords.xVel, y: coords.yVel }
  }
});

for (let i = 0; i < configuration.seconds; i++){
  bots.forEach(bot => moveBot(bot, configuration.xLim, configuration.yLim));
}

const quadrants = countBotsInQuadrants(bots, configuration.xLim, configuration.yLim);

console.log(quadrants.q1 * quadrants.q2 * quadrants.q3 * quadrants.q4);
// functions

function moveBot(bot: Bot, xLim: number, yLim: number): void {
  let newX = bot.pos.x + bot.velocity.x;
  let newY = bot.pos.y + bot.velocity.y;

  if (newX < 0){
    newX = newX + xLim;
  } else if (newX >= xLim){
    newX = newX - xLim;
  }

  if (newY < 0){
    newY = newY + yLim;
  } else if (newY >= yLim){
    newY = newY - yLim;
  }

  bot.pos = { x: newX, y: newY };
}

function countBotsInQuadrants(bots: Bot[], xLim: number, yLim: number): { q1: number, q2: number, q3: number, q4: number } {

  const quadrants = {
    q1: 0,
    q2: 0,
    q3: 0,
    q4: 0
  }

  const mids = {
    x: Math.floor(xLim / 2),
    y: Math.floor(yLim / 2)
  }

  for (const bot of bots){
    const x = bot.pos.x;
    const y = bot.pos.y;

    // skip robots on the middle lines
    if (x == mids.x || y == mids.y){
      continue;
    }

    if (x > mids.x && y > mids.y){
      quadrants.q1++;
    } else if (x < mids.x && y > mids.y){
      quadrants.q2++;
    } else if (x < mids.x && y < mids.y){
      quadrants.q3++;
    } else if (x > mids.x && y < mids.y){
      quadrants.q4++;
    }
  }

  return quadrants;
}

function getBotFromLine(line: string) {
  
  // match all numbers
  const items = line.matchAll(/-?\d+/g);

  if (!items){
    throw new Error('Invalid input');
  }

  const [x, y, xVel, yVel] = Array.from(items).map(Number);
  return { x, y, xVel, yVel };
}

